keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+either manually or as a library by other projects. It is safer to keep the old behavior of this
+command. However, we will deprecate it along with all the library functions. We can print a
+deprecation warning when this command is used, and promote the NodeMaintenance. Additionally, pods
+that support evacuation and have `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}`
+annotations will block eviction requests.
+
+`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
+the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
+As a consequence the node maintenance controller will reconcile the node back to the old value.
+Because of this we can make these commands noop when under the NodeMaintenance.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
+include Cordon and Drain toggles to support the following states/stages of the maintenance:
+1. Planning: this is to let the users know that maintenance will be performed on a particular set
+   of nodes in the future. Configured with `.spec.stage=Idle`.
+2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
+3. Drain: gives an intent to drain all selected nodes by creating `Evacuation` objects for the
+   node's pods. Configured with `.spec.stage=Drain`.
+4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
+   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
+   `Drained` condition is set to `"True"` on the node maintenance object.
+5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
+   Configured with `.spec.stage=Complete`.
+
+```golang
+
+// +enum
+type NodeMaintenanceStage string
+
+const (
+// Idle does not interact with the cluster.
+Idle NodeMaintenanceStage = "Idle"
+// Cordon cordons all selected nodes by making them unschedulable.
+Cordon NodeMaintenanceStage = "Cordon"
+// Drain:
+// 1. Cordons all selected nodes by making them unschedulable.
+// 2. Gives an intent to drain all selected nodes by creating Evacuation objects for the
+//    node's pods.
+Drain NodeMaintenanceStage = "Drain"
+// Complete:
+// 1. Removes all Evacuation objects requested by this NodeMaintenance.
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // NodeSelector selects nodes for this node maintenance.
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
```

Comment:
Be really careful with this language.  The way this is specced you have a state machine, which means any futre evolution will be very difficult.

Is it allowable for a cordon'ed node to become idle?  This state machine doesn't allow it.  I think you'd do better to describe this as a set of mutually exclusive intentions.  Modes.

Reply:
It is hard to offer this as modes, as it should be possible to manually switch between the states/stages. And also to have a strict transitions between them (please see [Supported Stage Transitions](https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#supported-stage-transitions)) as not all of them are valid. It is up to the admin or a controller when these transitions should occur. 

We also want to have a lifecycle for the NodeMaintenance. Once the NM completes we can clean up and recover the node. If an additional NM is required, a new object should be created.

I would be happy to hear ideas on how to improve this. 

If we decide for the current solution:
- It may be possible to add new stages into the API.
- The extensibility can be offered through a set of modifiers for each stage, as we do with the `DrainPlan`.